test/core/event_engine/test_suite/client_test.cc

```diff
@@ -12,9 +12,188 @@
+  static std::uniform_real_distribution<> dis(kMinMessageSize, kMaxMessageSize);
+  std::string tmp_s;
+  int len;
+  {
+    grpc_core::MutexLock lock(&g_mu);
+    len = dis(gen);
+  }
+  tmp_s.reserve(len);
+  for (int i = 0; i < len; ++i) {
+    tmp_s += alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return tmp_s;
+}
+
+}  // namespace
+
 // TODO(hork): establish meaningful tests
-TEST_F(EventEngineClientTest, TODO) {}
+
+// Create a connection using the test event engine to a non-existent listener
+// and verify that the connection fails.
+TEST_F(EventEngineClientTest, ConnectToNonExistentListenerTest) {
+  ConnectionManager mgr(this->NewEventEngine(), this->NewOracleEventEngine());
+  // Create a test event engine client endpoint and connect to a non existent
+  // oracle listener.
+  auto status = mgr.CreateConnection(
+      "ipv6:[::1]:7000", absl::InfiniteFuture(),
+      [](absl::Status status) {
+        // The On-Connect callback will be called with the status reported by
+        // the test event engine.
+        GPR_ASSERT(!status.ok());
+      },
+      false);
+  GPR_ASSERT(status.status() ==
+             absl::CancelledError("Failed to create connection."));
+}
+
+// Create a connection using the test event engine to a listener created
+// by the oracle event engine and exchange bi-di data over the connection.
+// For each data transfer, verify that data written at one end of the stream
+// equals data read at the other end of the stream.
+TEST_F(EventEngineClientTest, ConnectExchangeBidiDataTransferTest) {
+  ConnectionManager mgr(this->NewEventEngine(), this->NewOracleEventEngine());
+  std::string target_addr = "ipv6:[::1]:7000";
+  // Start an oracle localhost ipv6 listener
+  GPR_ASSERT(
+      mgr.StartListener(target_addr, /*listener_type_oracle=*/true).ok());
+  // Create a test event engine client endpoint and connect to oracle listener.
+  auto status = mgr.CreateConnection(
+      target_addr, absl::InfiniteFuture(),
+      [](absl::Status status) { GPR_ASSERT(status.ok()); }, false);
+  GPR_ASSERT(status.ok());
+  int connection_id = *status;
+  // Alternate message exchanges between client -- server and server -- client.
+  for (int i = 0; i < kNumExchangedMessages; i++) {
+    // Send from client to server and verify data read at the server.
+    GPR_ASSERT(mgr.TransferFromClient(/*connection_id=*/connection_id,
+                                      /*write_data=*/GetNextSendMessage())
+                   .ok());
+
+    // Send from server to client and verify data read at the client.
+    GPR_ASSERT(mgr.TransferFromServer(/*connection_id=*/connection_id,
+                                      /*write_data=*/GetNextSendMessage())
+                   .ok());
+  }
+  mgr.CloseConnection(connection_id);
+}
+
+// Create a N listeners and M connections where M > N and exchange and verify
+// data over each connection.
+TEST_F(EventEngineClientTest, MatrixOfConnectionsToOracleListenersTest) {
```

Comment:
I rewrote a sketch of this that doesn't rely on the ConnectionManager. I think it is more straightforward using the same EventEngine APIs that we hope to test directly. So again I don't see a great reason at this point to introduce a ConnectionManager.

https://gist.github.com/drfloob/31a50e7d1f296183623d6ed8a083b066

Reply:
I minimized the footprint of ConnectionMgr significantly now. It only supports binding and starting listeners and creating connections now. The connection manager API now returns unique_ptrs to client and server endpoints when the connection is successful.

I still feel a connection manager is required to remove some boiler plate code - particularly to retrieve the server endpoint. The server endpoint can be accessed only inside the listener's accept callback. So we need some way to block until the accept callback runs and then transfer the server endpoint to the test code.  At that point we also need to identify which connection the server endpoint maps to. Connection manager simplifies this stuff and returns both the client and server endpoint to the test to do as it pleases.

Reply:
I could change the connection mgr API if you think the test readability can still be improved

Reply:
I agree with you that the endpoint pair mapping logic might be duplicated a handful of places, but that's mostly in line with the recommendation to write DAMP tests. We chatted about this offline, but since I'm going to be on leave for a while, I'll catalog it briefly here for whomever gets involved next:

* prefer DAMP over DRY tests, even at the cost of some duplication, if it makes the test more readable. (See ToTT)
* a free function like `std::pair<Endpoint, Endpoint> CreateConnection(EventEngine* client_engine, EventEngine* listener_engine, std::vector<std::string> addrs)` could be used instead for the stated purpose of reducing boilerplate required to get a pair of connected endpoints, but I don't think that function would be very big (~20 lines maybe?), and if so, I think tests would likely be more readable if that code were inlined in each test. I provided a few test rewrites that elide the ConnectionManager, I think those are good examples of this.
* A method that has an argument of `bool create_type_a_or_b` is generally a design smell, which can likely be resolved with appropriate application of polymorphism.

Reply:
Cool. Per our offline discussion, modified client_test to remove ConnectionMgr usage. However, I've still kept the ConnectionMgr code around because I think it could be useful in the fuzzer implementation as well as for EE implementation specific tests.